noload: Add page-aligned padding to the ends of buffers

dlsloan opened the pull request:
ZFS often requests compress jobs which are not properly aligned for
nvme-algo in terms of buffer sizes. This patch pads the last part of
the source and dest buffers to guarantee page (4k) alignment of the
start of a buffer region and lba (512) alignment of the length for
nvme-algo.

lsgunth reviewed (changes requested):
This is very convoluted and does a lot of extra work in the common case where the size is aligned.

Can you also be clearer in the commit message about size aligned and what alignment is required... Which I thought was less then PAGE_SIZE, so theoretically we could tighten this up a bit...

Can we try something like this (not even compile tested) approach:

```diff
diff --git a/module/zfs/noload.c b/module/zfs/noload.c
index 3611b0e2ad55..83e4858ca9e8 100644
--- a/module/zfs/noload.c
+++ b/module/zfs/noload.c
@@ -73,7 +73,65 @@ void noload_release(void)
 		noload_disable();
 }
 
-static void bio_map_buf(struct bio *bio, void *data, unsigned int len)
+struct bio_pad_data {
+	void *orig, *bounce;
+	size_t len;
+};
+
+static void bio_copy_pad_endio(struct bio *bio)
+{
+	struct bio_pad_data *bpd = bio->bi_private;
+
+	memcpy(bpd->orig, bpd->bounce, len);
+	kfree(bpd->bounce);
+	kfree(bpd);
+	bio_put(bio);
+}
+
+static void bio_free_pad_endio(struct bio *bio)
+{
+	kfree(bio->bi_private);
+	bio_put(bio);
+}
+
+static int bio_bounce_pad(struct bio *bio, void *data, unsigned int len,
+			  bool is_dst)
+{
+	struct bio_pad_data *bpd;
+	void *bounce;
+
+	bounce = kmalloc(PAGE_SIZE, GFP_KERNEL);
+	if (!bounce)
+		return -ENOMEM;
+
+	if (is_dst) {
+		bpd = kmalloc(sizeof(*bpd), GFP_KERNEL);
+		if (!bpd) {
+			kfree(bounce);
+			return -ENOMEM;
+		}
+
+		bpd->orig = data;
+		bpd->bounce = bounce;
+		bpd->len = len;
+
+		bio->bi_private = bpd;
+		bio->bi_end_io = bio_copy_pad_endio;
+	} else {
+		memcpy(bounce, data, len);
+		memset(&bounce[len], 0, PAGE_SIZE - len);
+
+		bio->bi_private = bounce;
+		bio->bi_end_io = bio_free_pad_endio;
+	}
+
+	bio_add_page(bio, virt_to_page(bounce), PAGE_SIZE, 0);
+
+	return 0;
+}
+
+static int bio_map_buf(struct bio *bio, void *data, unsigned int len,
+		       bool is_dst)
 {
 	unsigned long kaddr = (unsigned long)data;
 	unsigned long end = (kaddr + len + PAGE_SIZE - 1) >> PAGE_SHIFT;
@@ -91,8 +149,10 @@ static void bio_map_buf(struct bio *bio, void *data, unsigned int len)
 		if (len <= 0)
 			break;
 
-		if (bytes > len)
+		if (bytes > len) {
+			return bio_bounce_pad(bio, data, len, is_dst);
 			bytes = len;
+		}
 
 		if (!is_vmalloc)
 			page = virt_to_page(data);
@@ -105,15 +165,15 @@ static void bio_map_buf(struct bio *bio, void *data, unsigned int len)
 		len -= bytes;
 		offset = 0;
 	}
+
+	return 0;
 }
 
 static int abd_to_bio_cb(void *buf, size_t size, void *priv)
 {
 	struct bio *bio = priv;
 
-	bio_map_buf(bio, buf, size);
-
-	return 0;
+	return bio_map_buf(bio, buf, size, false);
 }
 
 static ssize_t __noload_run(struct nvme_algo *alg, abd_t *src, void *dst,
@@ -142,9 +202,13 @@ static ssize_t __noload_run(struct nvme_algo *alg, abd_t *src, void *dst,
 	bio_src->bi_end_io = bio_put;
 	bio_dst->bi_end_io = bio_put;
 
-	abd_iterate_func(src, 0, s_len, abd_to_bio_cb, bio_src);
+	ret = abd_iterate_func(src, 0, s_len, abd_to_bio_cb, bio_src);
+	if (ret)
+		// TODO CLEANUP
 
-	bio_map_buf(bio_dst, dst, d_len);
+	ret = bio_map_buf(bio_dst, dst, d_len, true);
+	if (ret)
+		// TODO CLEANUP
 
 	ret = nvme_algo_run(alg, bio_src, s_len, bio_dst, &out_len);
 	if (ret) {

```

dlsloan commented:
Looks good, I probably should have left this one draft for a bit it's pretty heavy on the extra checks & BUG_ON's right now and is only one step removed from my initial debug code

dlsloan committed:
noload: Add page-aligned padding to the ends of buffers

ZFS often requests compress jobs which are not properly aligned for
nvme-algo in terms of buffer sizes. This patch pads the last part of
the source and dest buffers to guarantee page (4k) alignment of the
start of a buffer region and lba (512) alignment of the length for
nvme-algo.

lsgunth commented on module/zfs/noload.c:
```diff
@@ -37,6 +38,11 @@ int nvme_algo_run(struct nvme_algo *alg, struct bio *src,
 void nvme_algo_put(struct nvme_algo *alg);
 
+struct bio_pad_data {
+	void orig, bounce;
```
void *?... I didn't think raw void would even compile...

dlsloan replied:
I pushed a bit early, fixed this an a few others... seems void compiles to int

lsgunth replied:
Normally gcc would give you this error:

```
<source>:5:10: error: variable or field 'blah' declared void
    5 |     void blah;
      |          ^~~~
```

But ZFS must not have the compile strict enough...

lsgunth reviewed (commented):
I only saw one issue... but I'll give this another pass once it's fixed.

dlsloan commented:
This branch currently causes NVMe timeouts, working on fix

FIO script to run in zfs mount using noload device:

```
[Global]
    direct=0
    rw=randwrite
    blocksize_range=10k-1M
    blocksize_unaligned=1
    numjobs=128
    size=120k
    fallocate=none
[Writer]
```

lsgunth commented on module/zfs/noload.c:
```diff
@@ -95,29 +155,29 @@ static void bio_map_buf(struct bio *bio, void *data, unsigned int len)
 		else
 			page = vmalloc_to_page(data);
 
-		bio_add_page(bio, page, bytes, offset);
```
If bytes > len then it shouldn't get here.... So why min(bytes, len)?

dlsloan replied:
The condition is `bytes > len && !IS_ALIGNED(len, ALGO_ALIGN)`
